src/htm/encoders/SimHashDocumentEncoder.hpp

```diff
@@ -0,0 +1,290 @@
+   *
+   * Encoding is accomplished using SimHash, a Locality-Sensitive Hashing (LSH)
+   * algorithm from the world of nearest-neighbor document similarity search.
+   * As SDRs can be of any size, we use the SHA3 hashing algorithm with the
+   * SHAKE256 Extendible Output Function (XOF). We deviate slightly from the
+   * standard SimHash algorithm in order to achieve sparsity.
+   *
+   * In addition to document similarity, an option is provided to toggle if
+   * token similarity "near-spellings" (such as "cat" and "cats") will receieve
+   * similar encodings or not. Another option is provided for manging
+   * English-language token case in/sensitivity.
+   *
+   * @code
+   *    #include <htm/encoders/SimHashDocumentEncoder.hpp>
+   *    #include <htm/encoders/types/Sdr.hpp>
+   *
+   *    SimHashDocumentEncoderParameters params;
+   *    params.size = 400u;
+   *    params.activeBits = 21u;
+   *
+   *    SDR output({ params.size });
+   *    SimHashDocumentEncoder encoder(params);
+   *    encoder.encode({ "bravo", "delta", "echo" }, output)  // weights 1
+   *    encoder.encode({{ "bravo", 2 }, { "delta", 1 }, { "echo", 3 }}, output)
+   *
+   * @see BaseEncoder.hpp
+   * @see SimHashDocumentEncoder.cpp
+   * @see Serializable.hpp
+   */
+  class SimHashDocumentEncoder : public BaseEncoder<std::vector<std::string>> {
+  public:
+    /**
+     * Constructor
+     * @public
+     * @see BaseEncoder.hpp
+     */
+    SimHashDocumentEncoder() {};
+    SimHashDocumentEncoder(const SimHashDocumentEncoderParameters &parameters);
+
+    /**
+     * Initialize
+     * @public
+     * @see BaseEncoder.hpp
+     */
+    void initialize(const SimHashDocumentEncoderParameters &parameters);
+
+    // Public Params
+    const SimHashDocumentEncoderParameters &parameters = args_;
+
+    /**
+     * Encode (Main calling style)
+     *
+     * Each token/letter will be hashed with SHA3+SHAKE256 to get a binary
+     *  digest output of desired `size`. These vectors will be stored in a
+     *  matrix for the next step of processing.
+     * Weights are added in during hashing and simhashing.
+     * After the loop, we SimHash the matrix of hashes, resulting in an
+     *  output SDR.
+     * If param "caseSensitivity" is passed, we'll lowercase all token strings.
+     * If param "tokenSimilarity" is passed, we'll additionally loop through
+     *  all the letters in all the tokens.
+     *
+     * @param :input: Document token strings (with weights) to encode,
+     *  ex: {{ "what", 3 }, { "is", 1 }, { "up", 2 }}.
+     *  Documents can contain any number of tokens > 0. Token order in the
+     *  document is ignored and does not effect the output encoding.
+     * @param :output: Result SDR to fill with result output encoding.
+     *
+     * @public
+     * @see SimHashDocumentEncoder.hpp
+     * @see encode(const std::vector<std::string> input, SDR &output)
+     */
+    void encode(const std::map<std::string, UInt> input, SDR &output);
```

Comment:
nit: add `override` - this might be a reason why the Metrics(SDR) does not work. 
nit: `map&`?

Reply:
Blocked on both of those, it doesn't look like `encode()` is going to play well with several different typed calling signatures.  The class definition takes a type via template, which encode (and overrides) must match.  Instead of 3 calling signatures for `encode()`, I could split it into: `encodeString()`, `encodeList()`, `encodeMap()`, and alias `encode()` to the right place?

Reply:
>instead of 3 calling signatures for encode(), I could split it into: encodeString(), encodeList(), encodeMap(), and alias encode() to the right place? nstead of 3 calling signatures for encode(), I could split it into: encodeString(), encodeList(), encodeMap(), and alias encode() to the right place?

hmm, I see. What do you think of these 3 alternatives? 
- is that that much of a problem, or we leave it as-is now? (least work)
- would it help (and break too much existing) if `BaseEncoder` didn't define the type for the whole class, but for `virtual void encode<T>(T input, SDR& output)` ?  (most correct?) 
- could you have an arg in the constructor (`mode="string/map/list" `) and call BaseEncoder with required type based on that?

Reply:
I'm thinking leaving it as-is will be the least-weird solution. I'll make sure it's clean as possible. The only downside is a possible non-blocking issue with SdrMetrics.